src/Package/Impl/Sql/Publish/SqlPublishOptionsDialogViewModel.cs

```diff
@@ -0,0 +1,315 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Microsoft.Common.Core.Shell;
+using Microsoft.Common.Wpf;
+using Microsoft.Languages.Core.Settings;
+using Microsoft.R.Components.Application.Configuration;
+using Microsoft.VisualStudio.R.Package.ProjectSystem;
+using Microsoft.VisualStudio.R.Package.ProjectSystem.Configuration;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.R.Package.Sql.Publish {
+    internal sealed class SqlPublishOptionsDialogViewModel : BindableBase {
+        private readonly ICoreShell _coreShell;
+        private readonly IProjectSystemServices _pss;
+        private readonly IWritableSettingsStorage _settingsStorage;
+        private readonly IProjectConfigurationSettingsProvider _pcsp;
+
+        private bool _canPublish;
+        private bool _targetHasName;
+        private bool _generateTable;
+        private int _selectedTargetIndex;
+        private int _selectedTargetTypeIndex;
+        private int _selectedCodePlacementIndex;
+        private int _selectedQuoteTypeIndex;
+        private string _targetTooltip;
+
+        class DbConnectionData {
+            public string Name;
+            public string ConnectionString;
+        }
+
+        private IReadOnlyList<string> _targets = new List<string>();
+        private IReadOnlyList<string> _targetProjects = new List<string>();
+        private IReadOnlyList<DbConnectionData> _targetConnections = new List<DbConnectionData>();
+
+        public Task InitializationTask { get; private set; } = Task.CompletedTask;
+
+        /// <summary>
+        /// Target types: DACPAC, database, project
+        /// </summary>
+        public IReadOnlyList<string> TargetTypeNames { get; private set; }
+        public int SelectedTargetTypeIndex {
+            get { return _selectedTargetTypeIndex; }
+            set {
+                if (_selectedTargetTypeIndex != value) {
```

Comment:
The problem is this chain is that methods call things on IProjectSystemServices that all require main thread. Ex ```GetDbConnections``` calls 

```
            var configuredProject = hierarchy?.GetConfiguredProject();
```
which goes through ```IVsHierarchy``` that is not thread friendly. So up until 

```
await provider.OpenProjectSettingsAccessAsync
```
context has to be main thread